### PR TITLE
[GH-#65] Added implementation of point addition

### LINKED ISSUE #65: Fix generation of child public key from parent public key
[Here](https://bitcoin.stackexchange.com/questions/62533/key-derivation-in-hd-wallets-using-the-extended-private-key-vs-hardened-derivati) we can see how the child key derivation should work. Problem lies with the implementation of the Elliptic curve [Point addition](https://en.wikipedia.org/wiki/Elliptic_curve_point_multiplication#Point_addition)

More info on the Elliptic curve could be found [here](https://eng.paxos.com/blockchain-101-elliptic-curve-cryptography) for further help. The idea of taking the x and y points of the public key come from there.

So here I have some code, still it's not working as it should 
```elixir
serialized_pub_key = KeyPair.compress(parent_key)

    <<derived_key::256, child_chain::binary>> =
      :crypto.hmac(:sha512, key.chain_code,
        <<serialized_pub_key::binary, index::size(32)>>)

    {tweak, _} = :crypto.generate_key(:ecdh, :secp256k1, derived_key)

    ## MAGIC HAPPENS HERE !!!!

    <<_rest::8, x_par::256, y_par::256>> = parent_key
    <<_rest::8, x_tw::256, y_tw::256>> = tweak

    p = 115792089237316195423570985008687907853269984665640564039457584007908834671663
    h = (y_tw - y_par) / (x_tw - y_par)
    IO.inspect round(h)
    lambda = rem(h, p)

    x_r = rem((lambda*lambda - x_tw - x_par), p)
    y_r = rem((lambda*(x_tw - x_r) - y_tw), p)

    child_key = <<4::8, x_r::size(256), y_r::size(256)>> |> KeyPair.compress()
```

Java implementation could be found [here](https://github.com/bitsofproof/supernode/blob/1.1/api/src/main/java/com/bitsofproof/supernode/api/ExtendedKey.java#L304)

Ruby implementation could be found [here](https://github.com/GemHQ/money-tree/blob/master/lib/money-tree/node.rb#L93)

C implementation [here](https://github.com/bitcoin-core/secp256k1/blob/master/src/secp256k1.c#L477)

### PR DESCRIPTION
**Pull request by gspasov**
To do so an erlang nif library has been added to the project - [libsecp256k1](https://github.com/mbrix/libsecp256k1)
It uses the Bitcoin secp256k1 C implementation.

To make it work you need to install couple of things beforehand if you don't have them already:
```sudo apt-get install autoconf autogen```
```sudo apt-get install libtool```
```sudo apt-get install libgmp3-dev```

resolves #65

### CODE REVIEW

**Comment by thepiwo** [comment]
does child public key from parent public key derivation work now?

**Comment by gspasov** [comment]
@thepiwo Yes, the test for the derivation is passing.

File: test/child_key_derivation.ex [hunk 1]
```diff
@@ -177,17 +177,17 @@ test "bip32_vectors_1_btc" do
...
-  #    |> KeyPair.derive("m/0'/1/2'/2")
-  #    |> KeyPair.generate_master_key(:public)
-  #  public_child_key = public_master_key |> KeyPair.derive("m/1000000000")
+  test "public_derivation" do
+    seed = "000102030405060708090a0b0c0d0e0f" |> Base.decode16!(case: :mixed)
+    master_key = KeyPair.generate_master_key(seed, :seed)
+    public_master_key =
+      master_key
+      |> KeyPair.derive("M/0'/1/2'/2")
+    IO.inspect public_master_key
```
Comment (meivantodorov) [inline test/child_key_derivation.ex]: I do not think that we need `IO.inspect`. Aside that - it looks ok to me.

Comment (gspasov) [reply test/child_key_derivation.ex]: I'll fix it

**Comment by thepiwo** [comment]
@andonov can you merge?

**Pull request merged**